### PR TITLE
Update to AbstractMCMC 2

### PR DESCRIPTION
**Pull request by devmotion**
This is the Turing counterpart to https://github.com/TuringLang/DynamicPPL.jl/pull/150.

CI tests will fail as long as there is no new release of DynamicPPL with the changes from the linked PR, but tests pass locally with the branch in DynamicPPL.

### CODE REVIEW

**Review by cpfiffer — APPROVED** [review]

File: src/contrib/inference/dynamichmc.jl [hunk 1]
```diff
@@ -41,82 +41,83 @@ function DynamicNUTS{AD}(space::Symbol...) where AD
...
-mutable struct DynamicNUTSState{V<:VarInfo, D} <: AbstractSamplerState
+struct DynamicNUTSState{V<:AbstractVarInfo,D}
     vi::V
     draws::Vector{D}
 end
 
 DynamicPPL.getspace(::DynamicNUTS{<:Any, space}) where {space} = space
 
-function AbstractMCMC.sample_init!(
+# initial step: general interface for resuming and
```
Comment (cpfiffer) [inline src/contrib/inference/dynamichmc.jl:51]: Looks like this comment wasn't finished.

File: src/inference/Inference.jl [hunk 2]
```diff
@@ -439,95 +352,54 @@ function AbstractMCMC.bundle_samples(
...
-        for v in values(params)
-            push!(vs, v[1])
-        end
-
-        push!(k, :lp)
-
-        nts[i] = NamedTuple{tuple(k...)}(tuple(vs..., getlogp(t)))
+    return map(ts) do t
+        params = map(first, getparams(t))
+        return merge(params, (lp = getlogp(t),))
```
Comment (cpfiffer) [inline src/inference/Inference.jl:372]: Couldn't we use `metadata(t)` here?

**Review by cpfiffer — APPROVED** [review]
Crap, I reviewed the test files in the DynamicPPL PR instead. I replicated my comments from there over to here -- sorry for the noise.

I've only got two minor comments.

File: src/inference/Inference.jl (hunk 2 above)
Comment (devmotion) [reply src/inference/Inference.jl:372]: Sure, we can. I guess I didn't want to change the current behaviour but maybe it is actually better (and more consistent with the Chains output) to add the full metadata?

File: src/contrib/inference/dynamichmc.jl (hunk 1 above)
Comment (devmotion) [reply src/contrib/inference/dynamichmc.jl:51]:
```suggestion
```

File: src/inference/Inference.jl (hunk 2 above)
Comment (cpfiffer) [reply src/inference/Inference.jl:372]: Yeah, it'll prevent future errors I think if we start to lean more heavily on `metadata` (which is correct, IMO).

```suggestion
        return merge(params, metadata(t))
```

**Review by torfjelde — COMMENTED** [review]
From a brief look, this looks identical to the Turing-part in DynamicPPL, right?

**Comment by devmotion** [comment]
> From a brief look, this looks identical to the Turing-part in DynamicPPL, right?

Yes, it is the same apart from the commits that I added yesterday and today (i.e. basically your suggestions). I'll update the PR when a new version of DynamicPPL is released.

**Comment by torfjelde** [comment]
Cool, cool; I'll approve this once we have DPPL released then :+1:

**Comment by devmotion** [comment]
One thing that should be addressed is the documentation of the sampling interface. Both https://turing.ml/dev/docs/for-developers/interface and https://turing.ml/dev/docs/for-developers/how_turing_implements_abstractmcmc are outdated. Moreover, they feel a bit redundant and do not belong to Turing - the sampling interface is defined in AbstractMCMC and explained in the README there, and the special implementation is mostly DynamicPPL-specific (and not Turing).

Actually I think it might be more appropriate to move these longer explanations to DynamicPPL and AbstractMCMC (of course, after updating them) and maybe linking to them from the Turing docs. It would also be easier to update them, to reduce redundancy, and would avoid the impression that all these things only work or are designed for Turing. I also don't think that anyone who studies the DynamicPPL or AbstractMCMC repos currently will notice these more detailed explanations.

**Comment by devmotion** [comment]
Hmm seems there are some numerical inaccuracies on some platforms regarding the normalization of probability vectors of the Categorical distribution. Tests passed locally (Linux 64bit, Julia 1.5.3).

**Comment by torfjelde** [comment]
> Actually I think it might be more appropriate to move these longer explanations to DynamicPPL and AbstractMCMC (of course, after updating them) and maybe linking to them from the Turing docs. It would also be easier to update them, to reduce redundancy, and would avoid the impression that all these things only work or are designed for Turing. I also don't think that anyone who studies the DynamicPPL or AbstractMCMC repos currently will notice these more detailed explanations.

Given how many packages we have at this point, it's really a hassle to have one big centralized documentation like we do now. IMO we ought to have separate project documentation for each package, with the possible exception of Turing.jl which is then hosted on turing.ml, and then we simply link (or if we really want to, re-render them) on Turing.jl. This significantly lowers the barrier to actually document the packages we're working on. But yeah, this is a larger issue and should be addressed outside of this PR of course :sweat_smile:

**Review by torfjelde — APPROVED** [review]
Given that this is the same as for DynamicPPL, with the exception of the last commit improving the testing of `GibbsConditional` (which I looked at, and it looks good), this looks good to me:)

**Pull request merged**